working/static metaprogramming/intro.md

```diff
@@ -0,0 +1,431 @@
+# Static Metaprogramming in Dart
+
+## Problem Statement
+
+Programmers hate repetition—having to write the same or very similar code over
+and over again. It’s boring, tedious, error-prone, and harder to maintain in
+parallel correctly. A key property of what makes a programming language
+productive is its facilities for eliminating repetition.
+
+The humble loop is a simple example. Instead of:
+
+```dart
+list.add(1);
+list.add(2);
+list.add(3);
+```
+
+Which is asking for a copy/paste error, we’d rather write:
+
+```dart
+for (var i = 1; i <= 5; i++) {
+  list.add(i);
+}
+```
+
+A more powerful example is functions. If you need to build a Flutter widget
+containing three similar columns, would you rather maintain this:
+
+```dart
+class MyApp extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return Row(
+      children: [
+        Column(
+          mainAxisSize: MainAxisSize.min,
+          mainAxisAlignment: MainAxisAlignment.center,
+          children: [Text('CALL')],
+        ),
+        Column(
+          mainAxisSize: MainAxisSize.min,
+          mainAxisAlignment: MainAxisAlignment.center,
+          children: [Text('ROUTE')],
+        ),
+        Column(
+          mainAxisSize: MainAxisSize.min,
+          mainAxisAlignment: MainAxisAlignment.center,
+          children: [Text('SHARE')],
+        ),
+      ],
+    );
+  }
+}
+```
+
+Or this:
+
+```dart
+class MyApp extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return Row(
+      children: [
+        _buildColumn('CALL'),
+        _buildColumn('ROUTE'),
+        _buildColumn('SHARE'),
+      ],
+    );
+  }
+
+  Column _buildColumn(String label) {
+    return Column(
+      mainAxisSize: MainAxisSize.min,
+      mainAxisAlignment: MainAxisAlignment.center,
+      children: [Text(label)],
+    );
+  }
+}
+```
```

Comment:
ironically that's considered bad style in flutter... better to pull out the _buildColumn into its own widget class. :-)

Reply:
A problem this proposal would fix! Best of both worlds, devs can write the syntax that is familiar, but a pre-compiler can inline things, or make wrapper classes where it makes sense.